Add initial tests for OTM reverse proxy setup

hectcastro opened the pull request:
Number of conditions to test was growing fast enough to warrant a codified test suite. Usage details are outlined in the `README`.

hectcastro committed:
Add initial tests for OTM reverse proxy setup

Number of conditions to test was growing fast enough to warrant a
codified test suite. Usage details are outlined in the `README`.

hectcastro committed:
Ensure Leiningen plugin for Midje is in project.clj

jwalgran commented on README.md:
```diff
@@ -0,0 +1,17 @@
+
+## Usage
+
+To execute the test suite, first install [Leiningen](http://leiningen.org/), then execute the following command:
```
The leiningen install instructions do not mention that you need a JVM. For those totally new to Clojure, it would be a good idea to list that dependency here.

hectcastro replied:
Addressed in https://github.com/OpenTreeMap/otm-website-tests/commit/ab1a12bd2f2cbdc73fdb6fa003b4999249ee5ba5.

jwalgran commented on README.md:
````diff
@@ -0,0 +1,17 @@
+To execute the test suite, first install [Leiningen](http://leiningen.org/), then execute the following command:
+
+```bash
+$ lein midje
````
It would be helpful to document that you can use and env var to point the test suite at any host `OTM_BASE_URL=http:/somehost:8181 lein midje`

hectcastro replied:
Addressed in https://github.com/OpenTreeMap/otm-website-tests/commit/e632a44a6a31a8500ba7eab3658c0cbea424be37.

jwalgran commented on src/otm_website_tests/core.clj:
```diff
@@ -0,0 +1,12 @@
+  "Default options for clj-http."
+  {:follow-redirects false
+   :throw-exceptions false})
+
```
Adding a `get` helper to core removes some bolierplate noise from the facts and makes them more readable

``` clojure
(def get [relative-url]
  (client/get (str base-url relative-url) http-client-options))
```

```
(fact "it returns a not found from S3 if trailing slash exists"
    (let [resp (get "/css/")]
      (resp :status) => 404
      (s3-response? resp) => true))
```

hectcastro replied:
Addressed in https://github.com/OpenTreeMap/otm-website-tests/commit/4baacf421f8d9b93c094d3f63d56696037483de7.

jwalgran commented:
When I run this suite against the current production config, I get one failing test ("cssmap"), which is expected.

:+1:

hectcastro commented:
I mostly put this together as an easier way to keep track of the conditions we need to cover for the reverse proxy setup. There also wasn't a clear place to store the code other than here.

Before I address the comments, do we want to keep this project at all? I don't want to impose Clojure on the team just because it is what I picked. There may be some kind of HTTP acceptance testing framework that suites our needs better, but I didn't dig too far before switching to Clojure and an HTTP library.

jwalgran commented:
A black box acceptance test suite that can validate the integrated stack is a nice thing to have. It could be a natural home for performance regression tests.

Given the shape of our team, I think it makes sense for us to write low-ceremony, code-based acceptance tests. A framework is probably going to bring with it a DSL and/or a plain text parser that, given our lack of business analysts or a QA team, is probably just overhead.

A potential upside of running on the JVM might be running any JMeter suites we develop via an API rather then shelling out to another process.

I am biased, because I really enjoy Clojure and am comfortable with its vocabulary and tool chain, so I will argue against it.

If we stick with Python, we limit context switching and avoid bringing in another platform. A combination of [requests](http://www.python-requests.org/en/latest/) and [unittest](https://docs.python.org/2/library/unittest.html) could do the same things, with about the same volume of code. Something like [nose-watch](https://github.com/lukaszb/nose-watch) could probably replace the runner automation that `lein-midje` gives us.

Commit:
Include JVM dependency

Commit:
Document OTM_BASE_URL environment variable

Commit:
Use HTTP client helper for HEAD requests

hectcastro commented:
Addressed the comments, but still not clear which way to move forward. If we switch to Python, I'd probably keep this branch, but open a new PR using the Python test libraries.

jwalgran commented:
Nice updates. Using `head` requests is a good idea.

I have not heard any objections to merging this as-is. I say we roll with it.

:+1:
